Azure/msrestazure-for-python
# New LRO implementation

Initially, I wanted to fix the 200/PATCH/Location situation that must be detected as InProgress and not Finished as currently. Finally I went further into a refactor/learning experience. This PR to capitalize the work done these last two days and get first feedback from @annatisch and @brettcannon 

Major points on code:

- Several function renames to better fit actual content. Split following SRP pattern sometime.
- Fix for 200/PATCH/Location situation
- Algorithm rebuilt:

  - Resource is now replaced by each call. No more keeping first resource as backup (discussed with @annatisch)
  - Get status for each case Async/Location/Resource situation have been revamped based one the C# client

Unittests:

- Fix incorrect use of Headers in update mock
- Fix incomplete getoutputs mock. Now raise if model is incomplete, to better fit real situation. Related to https://github.com/Azure/msrest-for-python/pull/5
- Add 200/Patch/Location test (from DocDB)
- Add 200/Patch/Async test (inspired from DocDB)
- Add 201/PUT/Succeeded/No Headers (from DevTestLabs)

Travis might need https://github.com/Azure/msrest-for-python/pull/5 to be green, we will see, all tests green on my machine.

19 tests KO (for 46 OK) on the Azure SDK for Python tests. At first sight, it's mostly (more than half) because we need to record the tests again, since we're doing in some situation one final GET that we weren't doing before (because we were using the Resource from the initial requests). More investigation tomorrow.
